File: browser-switch/src/main/java/com/braintreepayments/browserswitch/BrowserSwitchRequest.java
package com.braintreepayments.browserswitch;

import android.net.Uri;

import org.json.JSONException;
import org.json.JSONObject;

class BrowserSwitchRequest {

    private Uri uri;
    private String state;
    final private int requestCode;
    private JSONObject metadata;

    static final String PENDING = "PENDING";
    static final String SUCCESS = "SUCCESS";

    static BrowserSwitchRequest fromJson(String json) throws JSONException {
        JSONObject jsonObject = new JSONObject(json);
        int requestCode = jsonObject.getInt("requestCode");
        String url = jsonObject.getString("url");
        String state = jsonObject.getString("state");
        JSONObject metadata = jsonObject.optJSONObject("metadata");
        return new BrowserSwitchRequest(requestCode, Uri.parse(url), state, metadata);
    }

    BrowserSwitchRequest(int requestCode, Uri uri, String state, JSONObject metadata) {
        this.uri = uri;
        this.state = state;
        this.requestCode = requestCode;
        this.metadata = metadata;
    }

    void setUri(Uri value) {
        uri = value;
    }

    Uri getUri() {
        return uri;
    }

    int getRequestCode() {
        return requestCode;
    }

    String getState() {
        return state;
    }

    public JSONObject getMetadata() {
        return metadata;
    }

    public void setMetadata(JSONObject metadata) {
        this.metadata = metadata;
    }

    void setState(String state) {
        this.state = state;
    }

    String toJson() throws JSONException {
        JSONObject result = new JSONObject();
        result.put("requestCode", requestCode);
        result.put("url", uri.toString());
        result.put("state", state);
        if (metadata != null) {
            result.put("metadata", metadata);
        }
        return result.toString();
    }
}

File: browser-switch/src/main/java/com/braintreepayments/api/BrowserSwitchRequest.kt
package com.braintreepayments.api

import android.net.Uri
import kotlinx.serialization.Serializable
import org.json.JSONException
import org.json.JSONObject

@Serializable(with = BrowserSwitchRequestSerializer::class)
class BrowserSwitchRequest internal constructor(
    val requestCode: Int,
    val url: Uri,
    val metadata: JSONObject?,
    private val returnUrlScheme: String,
    var shouldNotifyCancellation: Boolean
) {

    @Throws(JSONException::class)
    fun toJson(): String {
        val result = JSONObject()
        result.put("requestCode", requestCode)
        result.put("url", url.toString())
        result.put("returnUrlScheme", returnUrlScheme)
        result.put("shouldNotify", shouldNotifyCancellation)
        if (metadata != null) {
            result.put("metadata", metadata)
        }
        return result.toString()
    }

    fun matchesDeepLinkUrlScheme(url: Uri): Boolean {
        return url.scheme != null && url.scheme.equals(returnUrlScheme, ignoreCase = true)
    }

    companion object {
        @Throws(JSONException::class)
        fun fromJson(json: String?): BrowserSwitchRequest {
            val jsonObject = JSONObject(json)
            val requestCode = jsonObject.getInt("requestCode")
            val url = jsonObject.getString("url")
            val returnUrlScheme = jsonObject.getString("returnUrlScheme")
            val metadata = jsonObject.optJSONObject("metadata")
            val shouldNotify = jsonObject.optBoolean("shouldNotify", true)
            return BrowserSwitchRequest(
                requestCode,
                Uri.parse(url),
                metadata,
                returnUrlScheme,
                shouldNotify
            )
        }
    }
}